server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -383,6 +383,7 @@ private void snapshot(
             SnapshotIndexCommit snapshotIndexCommit = null;
             try {
                 snapshotIndexCommit = new SnapshotIndexCommit(indexShard.acquireIndexCommitForSnapshot());
+                final var shardStateId = getShardStateId(indexShard, snapshotIndexCommit.indexCommit()); // not aborted so indexCommit() ok
                 snapshotStatus.addAbortListener(makeAbortListener(indexShard.shardId(), snapshot, snapshotIndexCommit));
                 snapshotStatus.ensureNotAborted();
```

Comment:
I need some help to understand why this fix works. The change is to call `getShardStateId` before `snapshotStatus.addAbortListener` and the goal is to make sure `snapshotIndexCommit` is not released when the `indexCommit()` method is call against it.

It seems the fix would work if some how `snapshotStatus.addAbortListener(...)` immediately calls the added listener which in turn releases `snapshotIndexCommit`. This suggests `snapshotStatus` is already aborted. But the next line `snapshotStatus.ensureNotAborted()` says it must not be aborted already. So I am confused.

Reply:
The issue is that an abort may happen concurrently, in between calling `snapshotStatus.ensureNotAborted()` and `snapshotIndexCommit.indexCommit()`.

Reply:
Ah right. I always need a prompt for concurrency. Thanks!